betty/job/scheduler/default.py

```diff
@@ -0,0 +1,267 @@
+        done: Callable[[Iterable[str]], Awaitable[None]],
+        jobs: Sequence[Job[_ContextCoT]],
+    ):
+        self._scheduler = scheduler
+        self._done = done
+        self._jobs = jobs
+
+    async def __call__(self) -> None:
+        try:
+            for job in self._jobs:
+                await job(self._scheduler)
+        except BaseException as reason:
+            await self._scheduler.cancel(reason)
+            return
+        await self._done(job.id for job in self._jobs)
+
+
+@final
+class _UnknownJob:
+    def __new__(cls):  # pragma: no cover  # noqa D102
+        raise RuntimeError(f"The {cls} sentinel cannot be instantiated.")
+
+
+@final
+@threadsafe
+class DefaultScheduler(Generic[_ContextCoT], Scheduler[_ContextCoT]):
+    """
+    Betty's default job scheduler.
+    """
+
+    def __init__(self, context: _ContextCoT, *, progress: Progress | None = None):
+        self._context = context
+        self._progress = progress or NoOpProgress()
+        self._lock = AsynchronizedLock.new_threadsafe()
+        self._cancelled = False
+        self._cancelled_reason: BaseException | None = None
+        self._completed = False
+        self._jobs: MutableMapping[str, Job[_ContextCoT] | type[_UnknownJob]] = (
+            defaultdict(lambda: _UnknownJob)
+        )
+        self._job_dependencies: MutableMapping[str, set[str]] = defaultdict(set)
+        self._scheduled_jobs: set[str] = set()
+        self._available_jobs_queue: MutableSequence[str] = []
+        self._available_jobs_sorter: TopologicalSorter[str] | None = None
+        self._issued_jobs: set[str] = set()
+        self._done_jobs: set[str] = set()
+
+    @override
+    @property
+    def context(self) -> _ContextCoT:
+        return self._context
+
+    def _is_open(self) -> bool:
+        return not self._cancelled and not self._completed
+
+    def _assert_open(self) -> None:
+        self._assert_not_cancelled()
+        self._assert_not_completed()
+
+    def _assert_not_cancelled(self) -> None:
+        if self._cancelled:
+            if isinstance(self._cancelled_reason, Cancelled):
+                raise self._cancelled_reason
+            raise Cancelled from self._cancelled_reason
+
+    def _assert_not_completed(self) -> None:
+        if self._completed:
+            raise Completed
+
+    @override
+    async def add(self, *jobs: Job[_ContextCoT]) -> None:
+        async with self._lock:
+            self._assert_open()
+            self._available_jobs_sorter = None
+            for job in jobs:
+                if job.id in self._jobs:
+                    raise DuplicateJobError(job.id)
+                self._jobs[job.id] = job
+                self._scheduled_jobs.add(job.id)
+                self._job_dependencies[job.id].update(
+                    job.dependencies - self._done_jobs
+                )
+            await self._progress.add(len(jobs))
+
+    def _build_available_jobs_sorter(self) -> None:
```

Comment:
Merge this with the update method to ensure that when we create a new sorter (which may include already issued jobs) releases all ready jobs (including already issued ones) so they can then be marked done safely.